### PR TITLE
new packages: bitcoin and lnd

### PR DESCRIPTION
**Pull request by stefanwouldgo**

### CODE REVIEW

File: packages/bitcoin/build.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+TERMUX_PKG_HOMEPAGE=https://bitcoincore.org/ 
+TERMUX_PKG_DESCRIPTION="Bitcoin Core" 
+TERMUX_PKG_LICENSE="MIT" 
+TERMUX_PKG_VERSION=0.18.1 
+TERMUX_PKG_SRCURL=https://github.com/bitcoin/bitcoin/archive/v$TERMUX_PKG_VERSION.tar.gz
+TERMUX_PKG_SHA256=db3c76ac913adfd67e3c7ff243b443c9706f81dd382d1212875fefc2de1ea5ff 
+TERMUX_PKG_DEPENDS="libevent,libzmq,boost,termux-services"
+TERMUX_PKG_BUILD_DEPENDS="pkg-config,util-linux"
+TERMUX_PKG_BUILD_IN_SRC=1
```
Comment (xeffyr) [inline packages/bitcoin/build.sh]: TERMUX_PKG_BUILD_IN_SRC accepts only either `true` or `false`.

File: packages/bitcoin/build.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+TERMUX_PKG_BUILDDIR=$TERMUX_PKG_SRCDIR
+
+termux_step_patch_package() {
+    patch -p1 < $TERMUX_PKG_BUILDER_DIR/0001-android-patches.patch
```
Comment (xeffyr) [inline packages/bitcoin/build.sh]: `termux_step_patch_package()` shouldn't be overridden. Default one is enough.

File: packages/bitcoin/build.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
+TERMUX_PKG_HOMEPAGE=https://bitcoincore.org/ 
+TERMUX_PKG_DESCRIPTION="Bitcoin Core" 
+TERMUX_PKG_LICENSE="MIT" 
+TERMUX_PKG_VERSION=0.18.1 
+TERMUX_PKG_SRCURL=https://github.com/bitcoin/bitcoin/archive/v$TERMUX_PKG_VERSION.tar.gz
+TERMUX_PKG_SHA256=db3c76ac913adfd67e3c7ff243b443c9706f81dd382d1212875fefc2de1ea5ff 
+TERMUX_PKG_DEPENDS="libevent,libzmq,boost,termux-services"
+TERMUX_PKG_BUILD_DEPENDS="pkg-config,util-linux"
+TERMUX_PKG_BUILD_IN_SRC=1
+TERMUX_PKG_BUILDDIR=$TERMUX_PKG_SRCDIR
```
Comment (xeffyr) [inline packages/bitcoin/build.sh]: Do not specify `TERMUX_PKG_BUILDDIR=$TERMUX_PKG_SRCDIR`. manually. `TERMUX_PKG_BUILD_IN_SRC` is enough.

File: packages/bitcoin/build.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+termux_step_pre_configure() {
+    (cd depends && make HOST=$TERMUX_HOST_PLATFORM NO_QT=1 -j $TERMUX_MAKE_PROCESSES)
+    cd $TERMUX_PKG_SRCDIR
+    ./autogen.sh
+}
+
+termux_step_configure() {
+    ./configure PATH=$PATH:$TERMUX_PREFIX/bin --disable-dependency-tracking  --disable-rpath --disable-rpath-hack --host=$TERMUX_HOST_PLATFORM --prefix=/home/builder/.termux-build/bitcoin/src/depends/$TERMUX_HOST_PLATFORM  ac_cv_c_bigendian=no ac_cv_sys_file_offset_bits=$TERMUX_ARCH_BITS --disable-bench --enable-experimental-asm --disable-test --disable-man --without-libs --with-daemon --disable-maintainer-mode --disable-glibc-back-compat --disable-nls --enable-shared --enable-static --libexecdir=$TERMUX_PREFIX/libexec --bindir=$TERMUX_PREFIX/bin
```
Comment (xeffyr) [inline packages/bitcoin/build.sh]: Any reason to execute configure manually ?
Additional flags can be passed through `TERMUX_PKG_EXTRA_CONFIGURE_ARGS`.

File: packages/bitcoin/0001-android-patches.patch [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+index 5f622f8e6..89c619ef6 100644
+--- i/depends/packages/libevent.mk
++++ w/depends/packages/libevent.mk
+@@ -5,7 +5,7 @@ $(package)_file_name=release-$($(package)_version).tar.gz
+ $(package)_sha256_hash=316ddb401745ac5d222d7c529ef1eada12f58f6376a66c1118eee803cb70f83d
+ 
+ define $(package)_preprocess_cmds
+-  ./autogen.sh
++  ./autogen.sh && patch -p1 < /home/builder/termux-packages/packages/bitcoin/0001-fixup.patch
+ endef
```
Comment (xeffyr) [inline packages/bitcoin/0001-android-patches.patch]: 1. Patching shouldn't be done in makefiles.

2. Never rely on absolute paths in build environment.

Comment (xeffyr) [reply packages/bitcoin/0001-android-patches.patch]: Generally, you need to only create patch files and place them to package dir. Nothing more should be done as patches handled automatically.

File: packages/bitcoin/build.sh (hunk 1 above)
Comment (stefanwouldgo) [reply packages/bitcoin/build.sh]: thanks, fixed.

File: packages/bitcoin/build.sh (hunk 2 above)
Comment (stefanwouldgo) [reply packages/bitcoin/build.sh]: I'm not so sure about this. I have copied the - admittedly complex - patch strategy from upstream (https://github.com/greenaddress/bitcoin_ndk) and had assumed there is a reason for it. Simply leaving the patch files in the package dir doesn't work: 
```
1 out of 1 hunk FAILED -- saving rejects to file configure.ac.rej
The text leading up to this was:
--------------------------
|diff --git a/evutil_rand.c b/evutil_rand.c
|index 046a14b..3f0bf2c 100644
|--- a/evutil_rand.c
|+++ b/evutil_rand.c
--------------------------
```

File: packages/bitcoin/build.sh (hunk 4 above)
Comment (stefanwouldgo) [reply packages/bitcoin/build.sh]: yeah, adding the extra args to TERMUX_PKG_EXTRA_CONFIGURE_ARGS doesn't work. It gets me "the compiler cannot create executables."

File: packages/bitcoin/0001-android-patches.patch (hunk 5 above)
Comment (stefanwouldgo) [reply packages/bitcoin/0001-android-patches.patch]: Any suggestions of how to handle this instead? variables don't seem to work in patch files and the second patch does not get handled automatically.

File: packages/bitcoin/build.sh (hunk 2 above)
Comment (xeffyr) [reply packages/bitcoin/build.sh]: If `patch -p1` doesn't work, it will not work here anyway.

File: packages/bitcoin/build.sh (hunk 4 above)
Comment (xeffyr) [reply packages/bitcoin/build.sh]: Is config.log available ?

File: packages/bitcoin/0001-android-patches.patch (hunk 5 above)
Comment (xeffyr) [reply packages/bitcoin/0001-android-patches.patch]: All patches **must** be applied by termux-packages scripts and not makefiles.

File: packages/bitcoin/build.sh (hunk 2 above)
Comment (xeffyr) [reply packages/bitcoin/build.sh]: 0001-fixup.patch can be applied only with `-p0` and therefore should be modified.

File: packages/bitcoin/build.sh (hunk 3 above)
Comment (xeffyr) [reply packages/bitcoin/build.sh]: 0001-fixup.patch can't be applied to current source code. This patch is probably outdated.

**Comment by xeffyr** [comment]
I also noticed that it tries to download boost while compiling which shouldn't happen.
```
termux - building bitcoin for arch aarch64...
Fetching boost_1_64_0.tar.bz2 from https://dl.bintray.com/boostorg/release/1.64.0/source/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
 21 76.7M   21 16.1M    0     0  2259k      0  0:00:34  0:00:07  0:00:27 3026k^Cmake: *** Deleting file '/home/builder/.termux-build/bitcoin/src/depends/sources/download-stamps/.stamp_fetched-boost-boost_1_64_0.tar.bz2.hash'
```

File: packages/bitcoin/build.sh (hunk 4 above)
Comment (xeffyr) [reply packages/bitcoin/build.sh]: Can't reproduce "the compiler cannot create executables.".

**Comment by xeffyr** [comment]
For now finishes with
```
checking for boostlib >= 1.47.0... yes
checking whether the Boost::System library is available... yes
configure: error: Could not find a version of the boost_system library!
```
since our boost doesn't have libboost_system.

**Comment by stefanwouldgo** [comment]
It downloads and builds its own dependencies, because the ones in termux don't work (e.g., it needs a specific (old) version of bdb, it needs libevent_pthreads, etc.). I've tried reproducing the deps in termux before, but I get nontrivial bugs in the resulting binary. The binary this build is producing works fine.

**Comment by xeffyr** [comment]
Strange. On archlinux bitcoind works with latest versions somehow. Without downloaded deps.

**Comment by stefanwouldgo** [comment]
I assume the problem is either libevent_pthreads (which isn't in termux, so I changed that build to get in in) or libzmq (the build was complaining about the library having moved before). Anyway, I get this behavior: https://github.com/bitcoin/bitcoin/issues/16751

**Comment by stefanwouldgo** [comment]
> Strange. On archlinux bitcoind works with latest versions somehow. Without downloaded deps.
> libevent_pthreads isn't available due to how libevent was packaged:
> 
> ```
> TERMUX_PKG_RM_AFTER_INSTALL="bin/event_rpcgen.py lib/libevent_*"
> ```

Yeah, I tried to remove that rm. It compiles and runs, but I get the above runtime problems.

**Comment by stefanwouldgo** [comment]
Why did you merge it? The build doesn't work in this state. It fails at patching.

**Comment by stefanwouldgo** [comment]
sorry, had the old docker container still running. testing the new build now.

**Comment by stefanwouldgo** [comment]
Unfortunately, the build now merged suffers from the exact problems I described here: https://github.com/bitcoin/bitcoin/issues/16751

**Pull request merged**